🚧 AGP 9

SimonMarquis opened the pull request:
This neat website can be used to track the current issues with AGP 9: https://agp-status.frybits.com/agp-9.0.0/

---

- [ ] https://github.com/firebase/firebase-android-sdk/issues/7293
- [ ] [AGP 9.0.0-alpha03 produces circular dependency on KSP tasks](https://issuetracker.google.com/issues/442128557)
  fixed in KSP [2.2.20-2.0.3](https://github.com/google/ksp/releases/tag/2.2.20-2.0.3)
- [ ] https://github.com/gradle/android-cache-fix-gradle-plugin/pull/1886
  Temporary workaround with `android.newDsl=false`
  ```
  > Failed to apply plugin 'org.gradle.android.cache-fix'.
     > Could not get unknown property 'unitTestVariants' for object of type com.android.build.gradle.internal.dsl.ApplicationExtensionImpl$AgpDecorated.
  ```
- [ ] https://github.com/google/dagger/issues/4944
  Temporary workaround with `android.newDsl=false`

liutikas commented on build-logic/convention/build.gradle.kts:
```diff
@@ -37,7 +37,7 @@ kotlin {
 
 dependencies {
-    compileOnly(libs.android.gradleApiPlugin)
+    compileOnly(libs.android.gradlePlugin)
```
Can you share why we need to make this change?

SimonMarquis replied:
This is obviously not a required change, but I decided to remove (hide) the fact that is was an "api" version of the dependency since the "impl" is not supposed to be used in future releases. Keeping the distinction between the api and the impl visible in the dependency name is not that useful.

liutikas commented on gradle/libs.versions.toml:
```diff
@@ -34,7 +34,7 @@ coil = "2.7.0"
 firebaseBom = "33.7.0"
 firebaseCrashlyticsPlugin = "3.0.6"
-firebasePerfPlugin = "2.0.1"
+firebasePerfPlugin = "2.0.2"
```
Will handle this upgrade in https://github.com/android/nowinandroid/pull/1983

liutikas commented:
> release build type has isDebuggable forced to true, even when explicitly turned off.

Can you point me to how to repro this?

SimonMarquis commented:
I've not been able to reproduce it on my other projects, but on this one, simply build a `release` app with `gradlew :app:assembleDemoRelease`.

SimonMarquis commented:
I found the issue, fixed in 52b118a1e073c07c4b7843e3ca93ea70333e41a0.
`enableAndroidTestCoverage` forces the app to be debuggable, probably because of a new logic internal to AGP.

I've refactored the code to configure only the `debug` buildTypes:
https://github.com/SimonMarquis/nowinandroid/blob/52b118a1e073c07c4b7843e3ca93ea70333e41a0/build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/Jacoco.kt#L68-L71

liutikas commented on gradle.properties:
```diff
@@ -63,3 +63,5 @@ roborazzi.test.verify=true
 # https://issuetracker.google.com/issues/295039976
 android.injected.androidTest.leaveApksInstalledAfterRun=true
+
+android.newDsl=false
```
Add a comment on why we need to opt out linking to the hilt bug.

liutikas commented on sync/work/src/prod/kotlin/com/google/samples/apps/nowinandroid/sync/di/SyncModule.kt:
```diff
@@ -28,8 +28,10 @@ import dagger.Module
+import org.jetbrains.annotations.VisibleForTesting
 import javax.inject.Singleton
 
+@VisibleForTesting
```
This seems to be unrelated to AGP upgrade. Please undo

SimonMarquis commented on core/datastore/src/main/kotlin/com/google/samples/apps/nowinandroid/core/datastore/di/DataStoreModule.kt:
```diff
@@ -33,15 +33,17 @@ import dagger.hilt.android.qualifiers.ApplicationContext
     @Provides
     @Singleton
-    internal fun providesUserPreferencesDataStore(
+    fun providesUserPreferencesDataStore(
```
Yes, see this commit: https://github.com/android/nowinandroid/pull/1959/commits/fd939d5a48898f263c9e737fcd8428395932d8bd

SimonMarquis replied:
Probably due to mangling. Unfortunate I'm be afk for a few weeks and I won't be able to create a proper bug report until I'm back.

SimonMarquis commented on build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/AndroidCompose.kt:
```diff
@@ -28,12 +28,10 @@ import org.jetbrains.kotlin.compose.compiler.gradle.ComposeCompilerGradlePluginE
-        buildFeatures {
-            compose = true
-        }
+        buildFeatures.compose = true
```
Since the reviewer asked for minimal changes, I reverted the changes with `.apply`.

SimonMarquis commented on build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/NiaFlavor.kt:
```diff
@@ -20,20 +20,20 @@ enum class NiaFlavor(val dimension: FlavorDimension, val applicationIdSuffix: St
                     dimension = niaFlavor.dimension.name
                     flavorConfigurationBlock(this, niaFlavor)
-                    if (this@apply is ApplicationExtension && this is ApplicationProductFlavor) {
+                    if (this@flavors is ApplicationExtension && this is ApplicationProductFlavor) {
```
The extension was removed from AGP but kotlin-dsl provides one that fits: [`7839613` (#1959)](https://github.com/android/nowinandroid/pull/1959/commits/7839613a5b2fab6c16a0f1beb9ee2cb693d25816)
